pkg/retention/retention.go

```diff
@@ -416,3 +419,108 @@ func DeleteMetricsSegmentData(mmetaFile string, metricSegmentsToDelete map[strin
+		return
+	}
+
+	totalInodes := stat.Files
+	freeInodes := stat.Ffree
+	usedInodes := totalInodes - freeInodes
+
+	if totalInodes == 0 {
+		log.Errorf("doInodeBasedDeletion: Invalid total inodes count: %d", totalInodes)
+		return
+	}
+
+	currentUsagePercent := (usedInodes * 100) / totalInodes
+
+	log.Infof("doInodeBasedDeletion: Current inode usage: %v%% (%d used of %d total), Max allowed: %v%%, IngestNodeDir: %v",
+		currentUsagePercent, usedInodes, totalInodes, maxInodeUsagePercent, ingestNodeDir)
+
+	if currentUsagePercent <= maxInodeUsagePercent {
+		return
+	}
+
+	if deletionWarningCounter < MAXIMUM_WARNINGS_COUNT {
+		log.Warnf("Skipping inode-based deletion since try %d, Current usage: %v%%, Max allowed: %v%%",
+			deletionWarningCounter, currentUsagePercent, maxInodeUsagePercent)
+		return
+	}
+
+	// Get all segments sorted by time
+	allSegMetas := writer.ReadLocalSegmeta(false)
+	currentMetricsMeta := path.Join(ingestNodeDir, mmeta.MetricsMetaSuffix)
+	allMetricMetas, err := mmeta.ReadMetricsMeta(currentMetricsMeta)
+	if err != nil {
+		log.Errorf("doInodeBasedDeletion: Failed to get metric meta entries, filepath=%v, err: %v", currentMetricsMeta, err)
+		return
+	}
+
+	// Combine and sort all entries by time
+	allEntries := make([]interface{}, 0, len(allMetricMetas)+len(allSegMetas))
+	for i := range allMetricMetas {
+		allEntries = append(allEntries, allMetricMetas[i])
+	}
+	for i := range allSegMetas {
+		allEntries = append(allEntries, allSegMetas[i])
+	}
+
+	sort.Slice(allEntries, func(i, j int) bool {
+		var timeI uint64
+		if segMeta, ok := allEntries[i].(*structs.SegMeta); ok {
+			timeI = segMeta.LatestEpochMS
+		} else if metricMeta, ok := allEntries[i].(*structs.MetricsMeta); ok {
+			timeI = uint64(metricMeta.LatestEpochSec * 1000)
+		} else {
+			return false
+		}
+
+		var timeJ uint64
+		if segMeta, ok := allEntries[j].(*structs.SegMeta); ok {
+			timeJ = segMeta.LatestEpochMS
+		} else if metricMeta, ok := allEntries[j].(*structs.MetricsMeta); ok {
+			timeJ = uint64(metricMeta.LatestEpochSec * 1000)
+		} else {
+			return false
+		}
+		return timeI < timeJ
+	})
+
+	// Delete oldest segments until we get below the threshold
+	segmentsToDelete := make(map[string]*structs.SegMeta)
+	metricSegmentsToDelete := make(map[string]*structs.MetricsMeta)
+
+	for _, metaEntry := range allEntries {
+		// Check if we've deleted enough by getting fresh inode stats
+		err := syscall.Statfs(filepath.Clean(dataPath), &stat)
+		if err != nil {
+			log.Errorf("doInodeBasedDeletion: Failed to get updated inode stats: %v", err)
+			break
+		}
+
+		currentUsagePercent = ((stat.Files - stat.Ffree) * 100) / stat.Files
+		if currentUsagePercent <= maxInodeUsagePercent {
+			break
+		}
+
+		switch entry := metaEntry.(type) {
+		case *structs.MetricsMeta:
+			metricSegmentsToDelete[entry.MSegmentDir] = entry
+		case *structs.SegMeta:
+			segmentsToDelete[entry.SegmentKey] = entry
+		}
```

Comment:
We are not actually deleting any files here. We are just adding them to the map of segments we want to delete.
So the below line
```
// Check if we've deleted enough by getting fresh inode stats
		err := syscall.Statfs(filepath.Clean(dataPath), &stat)
```
Checking for the latest `inodes` state on each iteration of the loop might not be necessary? Let me know if I miss something here.